homeassistant/components/homekit/type_fans.py

```diff
@@ -0,0 +1,152 @@
+        self._flag[CHAR_ACTIVE] = True
+
+        service = SERVICE_TURN_ON if value == 1 else SERVICE_TURN_OFF
+        params = {ATTR_ENTITY_ID: self.entity_id}
+        self.hass.services.call(DOMAIN, service, params)
+
+    @debounce
+    def set_speed(self, value):
+        """Set speed if call came from HomeKit."""
+        _LOGGER.debug('%s: Set speed to %d', self.entity_id, value)
+        self._flag[CHAR_ROTATION_SPEED] = True
+        if value != 0:
+            speed = self.speed_list[int(round(value / self.min_step)) - 1]
+            params = {ATTR_ENTITY_ID: self.entity_id,
+                      ATTR_SPEED: speed}
+            self.hass.services.call(DOMAIN, SERVICE_TURN_ON, params)
+        else:
+            params = {ATTR_ENTITY_ID: self.entity_id}
+            self.hass.services.call(DOMAIN, SERVICE_TURN_OFF, params)
+
+    def set_direction(self, value):
+        """Set state if call came from HomeKit."""
+        _LOGGER.debug('%s: Set direction to %d', self.entity_id, value)
+        self._flag[CHAR_ROTATION_DIRECTION] = True
+        direction = DIRECTION_REVERSE if value == 1 else DIRECTION_FORWARD
+        params = {ATTR_ENTITY_ID: self.entity_id,
+                  ATTR_DIRECTION: direction}
+        self.hass.services.call(DOMAIN, SERVICE_SET_DIRECTION, params)
+
+    def set_oscillating(self, value):
+        """Set state if call came from HomeKit."""
+        _LOGGER.debug('%s: Set oscillating to %d', self.entity_id, value)
+        self._flag[CHAR_SWING_MODE] = True
+        oscillating = True if value == 1 else False
+        params = {ATTR_ENTITY_ID: self.entity_id,
+                  ATTR_OSCILLATING: oscillating}
+        self.hass.services.call(DOMAIN, SERVICE_OSCILLATE, params)
+
+    def update_state(self, new_state):
+        """Update fan after state change."""
+        # Handle State
+        state = new_state.state
+        if state in (STATE_ON, STATE_OFF):
+            hk_state = 1 if state == STATE_ON else 0
+            if not self._flag[CHAR_ACTIVE] and \
+                    self.char_active.value != hk_state:
+                self.char_active.set_value(hk_state)
+            self._flag[CHAR_ACTIVE] = False
+
+        # Handle Speed
+        if CHAR_ROTATION_SPEED in self.chars:
+            speed = new_state.attributes.get(ATTR_SPEED)
+            if not self._flag[CHAR_ROTATION_SPEED] and \
+                    len(self.speed_list) > 1 and isinstance(speed, str):
+                hk_speed = (self.speed_list.index(speed) + 1) * self.min_step
+                if self.char_speed.value != hk_speed:
+                    self.char_speed.set_value(hk_speed)
+            self._flag[CHAR_ROTATION_SPEED] = False
+
+        # Handle Direction
+        if CHAR_ROTATION_DIRECTION in self.chars:
+            direction = new_state.attributes.get(ATTR_DIRECTION)
+            if not self._flag[CHAR_ROTATION_DIRECTION] and \
+                    direction in (DIRECTION_FORWARD, DIRECTION_REVERSE):
```

Comment:
This currently doesn't work if set from the frontend. See https://github.com/home-assistant/home-assistant-polymer/issues/1158